gh-95271: Extract placeholders howto from sqlite3 tutorial

erlend-aasland opened the pull request:
Fixes gh-95271

erlend-aasland committed:
gh-95271: Extract placeholders howto from sqlite3 tutorial

erlend-aasland committed:
Syntax

CAM-Gerlach commented on Doc/library/sqlite3.rst:
```diff
@@ -1444,6 +1421,33 @@ Python types via :ref:`converters <sqlite3-converters>`.
 
+.. _sqlite3-placeholders:
+
+SQL operations usually need to use values from Python variables. However,
```
This is missing a section heading, e.g.

```suggestion
.. _sqlite3-placeholders:

Using placeholders in SQLite queries
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

SQL operations usually need to use values from Python variables. However,
```

If you prefer a different one, feel free to use it instead.

erlend-aasland replied:
🤦🏻‍♂️ 

Thanks!

erlend-aasland replied:
```suggestion
.. _sqlite3-placeholders:

Using placeholders to bind values in SQL queries
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

SQL operations usually need to use values from Python variables. However,
```

erlend-aasland replied:
... or _"Binding values to SQL statements using placeholders"_

CAM-Gerlach replied:
I find the first option a lot more clear, IMO

CAM-Gerlach reviewed (changes requested):
Given the stated scope of this PR I won't comment on anything but the move, but there is one pretty important thing missing—the section heading for the new How-To section :)

Otherwise, LGTM within the stated scope.

ezio-melotti commented on Doc/library/sqlite3.rst:
```diff
@@ -85,7 +85,9 @@ At this point, our database only contains one row::
-using :meth:`~Cursor.executemany`::
+using :meth:`~Cursor.executemany`.
+We use :ref:`question mark placeholders <sqlite3-placeholders>`
+to bind ``data`` to the SQL statement::
```
I think the part about using placeholders rather than string formatting was a pretty important thing that anyone using `sqlite3` should be aware of, and this move makes it a lot easier to miss.

It would be better to replace this sentence with a short paragraph -- possibly after the example -- that:
* points out that placeholders were used instead of string formatting;
* warns the user against using string formatting;
* explicitly links to the how-to (e.g. "For more information about placeholders, see :ref:`...`.")

erlend-aasland replied:
Yes, I think you are right; we should point out the danger of injection attacks. I have an idea of how to phrase it; let me sleep on it.

erlend-aasland replied:
CAM, Ezio: Please take a new look.

I think it worked out ok. I'm not sure about the string formatting link, but I think it is ok since this is a tutorial. I also added a wikipedia link to SQL injection attacks.

Commit:
Update Doc/library/sqlite3.rst

CAM-Gerlach reviewed (approved):
LGTM from me, thanks, aside from what @ezio-melotti pointed out

erlend-aasland committed:
Merge branch 'main' into sqlite-tutorial/extract-placeholders

erlend-aasland committed:
Address review: mention injection attacks

erlend-aasland committed:
Improvements:

- plural agreement, as CAM says
- more brevity
- add wikipedia link to SQL injection attacks

CAM-Gerlach commented on Doc/library/sqlite3.rst:
```diff
@@ -94,6 +94,12 @@ using :meth:`~Cursor.executemany`::
    ... ]
    >>> cur.executemany('INSERT INTO stocks VALUES(?, ?, ?, ?, ?)', data)
 
+Notice that we used question mark placeholders to bind *data* to the query.
```
```suggestion
Notice that we used ``?`` placeholders to bind *data* to the query.
```

May as well use the literal here, no? Its shorter, more explicit, and avoids any cross-language interpretation issues.

erlend-aasland replied:
Absolutely!

CAM-Gerlach reviewed (approved):
One minor comment, otherwise LGTM for now—thanks @erlend-aasland !

Commit:
Use literal question mark

Co-authored-by: CAM Gerlach <CAM.Gerlach@Gerlach.CAM>

ezio-melotti commented on Doc/library/sqlite3.rst:
```diff
@@ -94,6 +94,12 @@ using :meth:`~Cursor.executemany`::
    >>> cur.executemany('INSERT INTO stocks VALUES(?, ?, ?, ?, ?)', data)
 
+Notice that we used ``?`` placeholders to bind *data* to the query.
+Always use placeholders instead of :ref:`string formatting<tut-formatting>`
```
```suggestion
Always use placeholders instead of :ref:`string formatting <tut-formatting>`
```
Weird, `sphinx-lint` should have caught this :thinking: 

Edit: apparently this is valid markup -- I was confusing it with this checker: <https://github.com/sphinx-contrib/sphinx-lint/blob/main/sphinxlint.py#L345>.  I would suggest fixing it anyway for consistency.

erlend-aasland replied:
Yes, that valid, but I agree it looks weird. +1 for consistency!

ezio-melotti replied:
As I just mentioned in https://github.com/python/cpython/pull/95626#pullrequestreview-1062529930, there is at least another occurrence of this in the file.  You might want to check/fix them all in a separate PR.

erlend-aasland replied:
Thanks for the heads-up!

Commit:
Consistency FTW

Co-authored-by: Ezio Melotti <ezio.melotti@gmail.com>
